### ISSUE TITLE
Integer Inconsistencies

### ISSUE THREAD

**Issue by ssnepenthe** [pantheon-systems/wp-redis#181]
As far as I can tell, integers seem to be the only type where the value you put in is not guaranteed to be the same as the value you get back out.

It looks like the reason is that all values are serialized before save except for integers (I assume this is to accommodate `->incr()` and `->decr()`?). The problem is that Redis actually stores integers internally as strings.

```php
wp_cache_set( 'int', 6 );

wp_cache_get( 'int' ); // integer 6, because it is retrieved from the local cache
wp_cache_get( 'int', '', true ); // string '6', because it is retrieved from Redis
```

Is there any reason I might be missing why this behavior is actually desired? If not, I *think* the fix could be as simple as casting to int in the get method.

**Comment by danielbachhuber** [comment]
> As far as I can tell, integers seem to be the only type where the value you put in is not guaranteed to be the same as the value you get back out.

Related #96

> It looks like the reason is that all values are serialized before save except for integers (I assume this is to accommodate `->incr()` and `->decr()`?).

I _believe_ there wasn't an explicit decision for this.

The code history in WP Redis points to c8486576b0d55552570a678998a68e55a8ea1f56, which isn't helpful.

> Is there any reason I might be missing why this behavior is actually desired? If not, I _think_ the fix could be as simple as casting to int in the get method.

To my knowledge, no. @mboynes Do you recall?

I don't think casting as int in the get method is an appropriate solution because it runs the risk of casting data that wasn't meant to be cast.

The more correct solution would be to serialize integers on the way in, so we can be sure we persist type.

**Comment by ssnepenthe** [comment]
> I don't think casting as int in the get method is an appropriate solution because it runs the risk of casting data that wasn't meant to be cast.

I was thinking updates to lines 630 - 633:

```php
// All non-numeric values are serialized
$value = is_numeric( $value ) ? intval( $value ) : unserialize( $value );
```

I might be missing something, but since the only things going in unserialized are integers, it should be safe to assume anything that is numeric was originally provided as an integer.

The exception might be if you are bypassing the object cache and setting data manually in redis. However object cache keys look like they include a salt, prefix and group so I'm not sure this is something worth worrying about.

> The more correct solution would be to serialize integers on the way in, so we can be sure we persist type.

The problem with saving serialized integers would be that you lose the ability to call the native incr and decr redis commands.

Not the end of the world - it looks like it is already being handled manually for non persistent groups. But it would also require updates to at least the incr, decr and _call_redis methods.

**Comment by ssnepenthe** [comment]
I just read over #96 - it looks like null is converted to an empty string for the internal cache, but null is still serialized before being saved in redis.

So at the point where the value would be cast to int, a null value would be serialized and unaffected.

**Comment by danielbachhuber** [comment]
> I might be missing something, but since the only things going in unserialized are integers, it should be safe to assume anything that is numeric was originally provided as an integer.

From `is_numeric()` [documentation](http://php.net/manual/en/function.is-numeric.php):

> Finds whether the given variable is numeric. Numeric strings consist of optional sign, any number of digits, optional decimal part and optional exponential part. Thus +0123.45e6 is a valid numeric value. Hexadecimal (e.g. 0xf4c3b00c) and binary (e.g. 0b10100111001) notation is not allowed.

So, this isn't a safe assumption.

> The problem with saving serialized integers would be that you lose the ability to call the native incr and decr redis commands.

Oh, true. This must be why it was implemented this way in the first place.

Unless I'm missing something, it doesn't appear we can safely change the behavior. If we can't safely change the behavior, then this is an idiosyncrasy we'll need to live with.

Did this manifest as a bug in the real world, or just an implementation detail you ran into?

**Comment by ssnepenthe** [comment]
> ...it doesn't appear we can safely change the behavior.

I am still not totally convinced based on those examples...

Keep in mind that the is_numeric check would occur after fetching from redis but before unserializing. This means that if I save a numeric string or a float, it will be checking `is_numeric( $serialized_string )` which will return false.

Using the examples from the docs:

+0123.45e6 is numeric, but evaluates as float/double. This means it is serialized and would be unaffected by the change.
'+0123.45e6' is numeric, but being a string, it gets serialized and would be unaffected by the change.

0xf4c3b00c and 0b10100111001 are numeric but are also immediately interpreted as integers (4106465292 and 1337) and therefore saved as strings ("4106465292" and "1337").

If you attempt to retrieve either of these immediately you get back an integer, and if you retrieve them in a future request you get a string.

The proposed change would just ensure that you always get back an integer.

> Did this manifest as a bug in the real world, or just an implementation detail you ran into?

Just an implementation detail I noticed in one of my tests when I switched from the core object cache to redis.

I think PHP is lenient enough with conversion between integer and numeric strings that it is unlikely to cause any bugs.
